Objective-C/CBLCollection.mm

```diff
@@ -17,49 +17,128 @@
 //  limitations under the License.
 //
 
-#import "CBLCollection.h"
-#import "CBLIndexable.h"
 #import "CBLChangeListenerToken.h"
+#import "CBLCollection+Internal.h"
 #import "CBLCollectionChangeObservable.h"
+#import "CBLCoreBridge.h"
+#import "CBLDatabase+Internal.h"
+#import "CBLIndexable.h"
+#import "CBLIndexConfiguration+Internal.h"
+#import "CBLScope.h"
+#import "CBLScope+Internal.h"
+#import "CBLStatus.h"
+#import "CBLStringBytes.h"
+
+using namespace fleece;
 
 NSString* const kCBLDefaultCollectionName = @"_default";
 
-@implementation CBLCollection
+@implementation CBLCollection {
+    C4Collection* _c4col;
+}
 
-@synthesize count=_count, name=_name, scope=_scope;
+@synthesize count=_count, name=_name, scope=_scope, db=_db;
 
-- (instancetype) initWithName: (NSString*)name
-                        scope: (nullable CBLScope*)scope
-                        error: (NSError**)error {
-    CBLAssertNotNil(name);
-    
+- (instancetype) initWithDB: (CBLDatabase*)db
+               c4collection: (C4Collection*)c4collection {
+    CBLAssertNotNil(db);
+    CBLAssertNotNil(c4collection);
     self = [super init];
     if (self) {
-        _name = name;
-        if (scope)
-            _scope = scope;
+        C4CollectionSpec spec = c4coll_getSpec(c4collection);
+        
+        _db = db;
+        _c4col = c4coll_retain(c4collection);
+        _name = slice2string(spec.name);
+        _scope = [[CBLScope alloc] initWithDB: db name: slice2string(spec.scope)];
+        
+        CBLLogVerbose(Database, @"%@ Creating collection:%@ db=%@ scope=%@",
+                      self, _name, db, _scope);
     }
+    
     return self;
 }
 
+- (void) dealloc {
+    c4coll_release(_c4col);
+}
+
+- (NSString*) description {
+    return [NSString stringWithFormat: @"%@[%@.%@]%@", self.class, _scope.name, _name, _db];
+}
+
+#pragma mark - Indexable
+
 - (BOOL) createIndexWithName: (NSString*)name
                       config: (CBLIndexConfiguration *)config
                        error: (NSError**)error {
-    // TODO: add implementation
-    return NO;
+    CBLAssertNotNil(name);
+    CBLAssertNotNil(config);
+    
+    CBLDatabase* db = _db;
+    CBL_LOCK(db) {
```

Comment:
I think it is probably time to change the way we do the lock but let's finish CBLCollection and CBLScope first.

I think we could switch to use NSLock / NSRecursiveLock. Another option is to use LiteCore's AccessLock.

Reply:
https://issues.couchbase.com/browse/CBL-3308